workflow/artifacts/s3/s3.go

```diff
@@ -51,107 +56,126 @@ func (s3Driver *ArtifactDriver) Load(inputArtifact *wfv1.Artifact, path string)
-					ObjectLocking: outputArtifact.S3.CreateBucketIfNotPresent.ObjectLocking,
-				})
-				if err != nil {
-					log.Warnf("Failed to create bucket: %v. Error: %v", outputArtifact.S3.Bucket, err)
-				}
-			}
+// saveS3Artifact uploads artifacts to an S3 compliant storage
+// returns true if the upload is completed or can't be retried (non-transient error)
+// returns false if it can be retried (transient error)
+func saveS3Artifact(s3cli argos3.S3Client, path string, outputArtifact *wfv1.Artifact) (bool, error) {
+	isDir, err := file.IsDirectory(path)
+	if err != nil {
+		return true, fmt.Errorf("failed to test if %s is a directory: %v", path, err)
+	}
 
-			if isDir {
-				if err = s3cli.PutDirectory(outputArtifact.S3.Bucket, outputArtifact.S3.Key, path); err != nil {
-					log.Warnf("Failed to put directory: %v", err)
-					return false, nil
-				}
-			} else {
-				if err = s3cli.PutFile(outputArtifact.S3.Bucket, outputArtifact.S3.Key, path); err != nil {
-					log.Warnf("Failed to put file: %v", err)
-					return false, nil
-				}
-			}
-			return true, nil
+	createBucketIfNotPresent := outputArtifact.S3.CreateBucketIfNotPresent
+	if createBucketIfNotPresent != nil {
+		log.Infof("Trying to create bucket: %s", outputArtifact.S3.Bucket)
+		err := s3cli.MakeBucket(outputArtifact.S3.Bucket, minio.MakeBucketOptions{
+			Region:        outputArtifact.S3.Region,
+			ObjectLocking: outputArtifact.S3.CreateBucketIfNotPresent.ObjectLocking,
 		})
-	return err
+		if err != nil {
+			return !isTransientS3Err(err), fmt.Errorf("failed to create bucket %s: %v", outputArtifact.S3.Bucket, err)
+		}
+	}
+
+	if isDir {
+		if err = s3cli.PutDirectory(outputArtifact.S3.Bucket, outputArtifact.S3.Key, path); err != nil {
+			return !isTransientS3Err(err), fmt.Errorf("failed to put directory: %v", err)
+		}
+	} else {
+		if err = s3cli.PutFile(outputArtifact.S3.Bucket, outputArtifact.S3.Key, path); err != nil {
+			return !isTransientS3Err(err), fmt.Errorf("failed to put file: %v", err)
+		}
+	}
+	return true, nil
 }
 
 func (s3Driver *ArtifactDriver) ListObjects(artifact *wfv1.Artifact) ([]string, error) {
 	ctx, cancel := context.WithCancel(context.Background())
 	defer cancel()
 
 	var files []string
-	err := wait.ExponentialBackoff(wait.Backoff{Duration: time.Second * 2, Factor: 2.0, Steps: 5, Jitter: 0.1},
+	err := waitutil.Backoff(defaultRetry,
 		func() (bool, error) {
 			s3cli, err := s3Driver.newS3Client(ctx)
 			if err != nil {
-				return false, err
+				return !isTransientS3Err(err), fmt.Errorf("failed to create new S3 client: %v", err)
 			}
 			files, err = s3cli.ListDirectory(artifact.S3.Bucket, artifact.S3.Key)
 			if err != nil {
-				return false, err
+				return !isTransientS3Err(err), fmt.Errorf("failed to list directory: %v", err)
 			}
 			return true, nil
 		})
 
 	return files, err
 }
+
+// isTransientS3Err checks if an minio.ErrorResponse error is transient (retryable)
+func isTransientS3Err(err error) bool {
```

Comment:
minor - please put in `workflow/artifacts/s3/error_codes.go` (maybe rename that to `workflow/artifacts/s3/errors.go`)

Reply:
thanks! refactored in 513a4b5fb50aad5471ff22027538b41bb722397d